Send deterministic tool description and tell the model about the approval mode with each turn.

aibrahim-oai opened the pull request:
Send deterministic tool description -> avoid breaking the cache
Tell the model about the approval mode with each turn -> to keep it aware of its environment.

aibrahim-oai committed:
fmt + clippy: codex-core deterministic shell tool tests, conflict cleanup

dylan-hurd-oai commented on codex-rs/core/src/openai_tools.rs:
```diff
@@ -98,9 +95,7 @@ impl ToolsConfig {
         if matches!(approval_policy, AskForApproval::OnRequest) && !use_streamable_shell_tool {
-            shell_type = ConfigShellToolType::ShellWithRequest {
-                sandbox_policy: sandbox_policy.clone(),
-            }
```
I think we'll still have the same switching issue, unless we use ShellWithRequest regardless of approval type?

dylan-hurd-oai reviewed (commented):
Let's clean this up a bit and get CI passing, but I think this is a net improvement! We can solve the approval modes <> tool consistency issue on top of this.

dylan-hurd-oai reviewed (commented):
@aibrahim-oai looks like we might need to rebase after https://github.com/openai/codex/pull/3069 - we can probably wait until the follow-up is done to move to EnvironmentContext, and then land this

dylan-hurd-oai commented on codex-rs/core/src/openai_tools.rs:
```diff
@@ -200,7 +195,50 @@ fn create_shell_tool() -> OpenAiTool {
+GUIDANCE
+- If unsure whether a command will be blocked, prefer setting `with_escalated_permissions=true` with a clear justification.
+- The effective policy (ReadOnly / WorkspaceWrite{network_access} / DangerFullAccess) is provided separately; the above reference tells you how each policy behaves.
+"#;
```
@aibrahim-oai This does indeed seem clearer, but it is quite different from our existing description, and reverts some of our recent fixes, e.g. `- Reading files outside the current working directory (CWD).` It also has a different format - why the drastic change?

dylan-hurd-oai replied:
Given that much of this is covered in our [prompt](https://github.com/openai/codex/blob/main/codex-rs/core/prompt.md?plain=1#L149-L182), I wonder if we want to simplify our tool definition altogether, instead of repeating instructions

aibrahim-oai commented on codex-rs/core/src/openai_tools.rs:
```diff
@@ -200,7 +195,50 @@ fn create_shell_tool() -> OpenAiTool {
+The shell tool executes shell commands inside a Landlock-style sandbox. Some operations may require escalated permissions depending on the *effective sandbox policy* (communicated separately at runtime).
+
+POLICY REFERENCE (names + descriptions)
+- SandboxPolicy::ReadOnly
```
good catch

aibrahim-oai committed:
patch-tools

dylan-hurd-oai committed:
fix tests

aibrahim-oai committed:
add tests

aibrahim-oai committed:
add tests

aibrahim-oai committed:
patch-tools

aibrahim-oai committed:
patch-tools

aibrahim-oai committed:
patch-tools

aibrahim-oai committed:
patch-tools

aibrahim-oai committed:
patch-tools

aibrahim-oai committed:
patch-tools

aibrahim-oai committed:
patch-tools

aibrahim-oai committed:
patch-tools

aibrahim-oai committed:
patch-tools

aibrahim-oai committed:
clippy

aibrahim-oai committed:
add tests

aibrahim-oai committed:
clippy

aibrahim-oai committed:
clippy

aibrahim-oai committed:
add tests

aibrahim-oai committed:
squash: 19 commit(s) since origin/main

- fmt + clippy: codex-core deterministic shell tool tests, conflict cleanup
- patch-tools
- fix tests
- add tests
- add tests
- patch-tools
- patch-tools
- patch-tools
- patch-tools
- patch-tools
- patch-tools
- patch-tools
- patch-tools
- patch-tools
- clippy
- add tests
- clippy
- clippy
- add tests

aibrahim-oai committed:
Merge branch 'main' into patch-tools

aibrahim-oai committed:
Merge branch 'patch-tools' of https://github.com/openai/codex into patch-tools

aibrahim-oai committed:
rebase

aibrahim-oai committed:
tests

aibrahim-oai committed:
tests
